CCDB-4518: Sanitize field names when converting to BQ schema

ncliang opened the pull request:
When the `sanitizeFieldNames` flag is set, the user is expecting the field names to be sanitized to conform to BigQuery naming convention. This is correctly done on the regular write path. But, if the user also specifies `autoCreateTables` , the schema of the table that is attempted to be created is not sanitized. This causes the connector to fail with the following exception
```
Caused by: com.wepay.kafka.connect.bigquery.exception.BigQueryConnectException: A write thread has failed with an unrecoverable error
Caused by: Failed to create table GenericData{classInfo=[datasetId, projectId, tableId], {datasetId=pgreprovincent, tableId=a_topic}}
        at com.wepay.kafka.connect.bigquery.write.batch.KCBQThreadPoolExecutor.lambda$maybeThrowEncounteredError$0(KCBQThreadPoolExecutor.java:101)
        at java.base/java.util.Optional.ifPresent(Optional.java:183)
        at com.wepay.kafka.connect.bigquery.write.batch.KCBQThreadPoolExecutor.maybeThrowEncounteredError(KCBQThreadPoolExecutor.java:100)
        at com.wepay.kafka.connect.bigquery.BigQuerySinkTask.put(BigQuerySinkTask.java:236)
        at org.apache.kafka.connect.runtime.WorkerSinkTask.deliverMessages(WorkerSinkTask.java:604)
        ... 10 more
Caused by: com.wepay.kafka.connect.bigquery.exception.BigQueryConnectException: Failed to create table GenericData{classInfo=[datasetId, projectId, tableId], {datasetId=pgreprovincent, tableId=a_topic}}
Caused by: Invalid field name "product 1". Fields must contain only letters, numbers, and underscores, start with a letter or underscore, and be at most 300 characters long.
        at com.wepay.kafka.connect.bigquery.write.row.AdaptiveBigQueryWriter.attemptTableCreate(AdaptiveBigQueryWriter.java:163)
        at com.wepay.kafka.connect.bigquery.write.row.AdaptiveBigQueryWriter.performWriteRequest(AdaptiveBigQueryWriter.java:102)
        at com.wepay.kafka.connect.bigquery.write.row.BigQueryWriter.writeRows(BigQueryWriter.java:112)
        at com.wepay.kafka.connect.bigquery.write.batch.TableWriter.run(TableWriter.java:93)
        ... 3 more
Caused by: com.google.cloud.bigquery.BigQueryException: Invalid field name "product 1". Fields must contain only letters, numbers, and underscores, start with a letter or underscore, and be at most 300 characters long.
        at com.google.cloud.bigquery.spi.v2.HttpBigQueryRpc.translate(HttpBigQueryRpc.java:113)
        at com.google.cloud.bigquery.spi.v2.HttpBigQueryRpc.create(HttpBigQueryRpc.java:186)
        at com.google.cloud.bigquery.BigQueryImpl$2.call(BigQueryImpl.java:253)
        at com.google.cloud.bigquery.BigQueryImpl$2.call(BigQueryImpl.java:250)
```

This PR pipes through `sanitizeFieldNames` to `SchemaManager` and `BigQuerySchemaConverter` so that they can sanitize the field names accordingly if requested to do so.

Unit and integration tests are added to verify field name sanitization behavior. The integration tests in `BigQuerySinkConnectorIT` are hard-coded to use avro, which itself follows the same naming convention as BigQuery. Simply adding a testcase to the suite would be insufficient to reproduce the invalid field name. Thus, a new IT class is added.

The PR is targeted as a bugfix release of the `2.0.x` branch.

C0urante commented on kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/convert/KafkaDataBuilder.java:
```diff
@@ -52,7 +53,9 @@ public static Field buildKafkaDataField(String kafkaDataFieldName) {
 
-        return Field.newBuilder(kafkaDataFieldName, LegacySQLTypeName.RECORD,
+        return Field.newBuilder(
+                FieldNameSanitizer.sanitizeName(kafkaDataFieldName),
```
Why are we unconditionally sanitizing here but conditionally sanitizing the key field in `SchemaManager::getRegularSchemaFields`?

Also, could we maybe move this logic into `SchemaManager:getRegularSchemaFields` just to avoid spreading it across so many different places and implementing it in slightly different ways?

ncliang replied:
Ah, good catch. Yes, moving this logic out into `getRegularSchemaFields` would consolidate the logic there a bit more. Will do.

C0urante commented on kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/SchemaManager.java:
```diff
@@ -66,6 +67,7 @@
   private final boolean allowNewBQFields;
   private final boolean allowBQRequiredFieldRelaxation;
   private final boolean allowSchemaUnionization;
+  private final boolean sanitizeFieldName;
```
Nit:
```suggestion
  private final boolean sanitizeFieldNames;
```

C0urante commented on kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/SchemaManager.java:
```diff
@@ -83,6 +85,7 @@
    * @param allowNewBQFields If set to true, allows new fields to be added to BigQuery Schema.
    * @param allowBQRequiredFieldRelaxation If set to true, allows changing field mode from REQUIRED to NULLABLE
    * @param allowSchemaUnionization If set to true, allows existing and new schemas to be unionized
+   * @param sanitizeFieldName If true, sanitizes field names on converted BigQuery schema
```
Nit:
```suggestion
   * @param sanitizeFieldNames If true, sanitizes field names to adhere to BigQuery column name restrictions
```

C0urante commented on kcbq-connector/src/test/java/com/wepay/kafka/connect/bigquery/integration/SanitizeFieldNameIT.java:
```diff
@@ -0,0 +1,171 @@
+  private static final String CONNECTOR_NAME = "kcbq-sink-connector";
+  private static final long NUM_RECORDS_PRODUCED = 20;
+  private static final int TASKS_MAX = 3;
+  private static final String KAFKA_FIELD_NAME = "kafkaKey";
```
This could be interpreted as referring to the Kafka data field, or the Kafka key field.
```suggestion
  private static final String KAFKA_KEY_FIELD_NAME = "kafkaKey";
```

C0urante reviewed (commented):
Thanks Nigel, this looks pretty good. Some nits and a few minor points in the comments.

Few thoughts here as well:
- Shouldn't we flesh out the `SchemaManagerTest` cases to cover the logic we use for sanitizing key (and possibly Kafka data) field names?
- Is an integration test really necessary? It looks like we're not adding any new coverage with the test case that isn't already addressed (or easier to address) with our tests, and given that the `SchemaManager` class is our single point of entry for table creation, it seems like fleshing out tests for that class might be more useful.
- Can we add test cases for when upsert/delete is enabled?

ncliang commented:
Thanks for review @C0urante . I had added the new IT as a end-to-end sanity check to verify the field name sanitation on the create table path. My first attempt was to add it as just a testcase to the `BigQuerySinkConnectorIT` , but then realized that the suite is hardcoded to use avro converter. The avro naming conventions prevented the testcase from specifying a field name that would be sanitized. 

Other than that, I think I've address your concerns. PTAL

C0urante commented:
Thanks Nigel. I think my point about the integration test is less that a new class isn't warranted (your assessment on that front seems reasonable) and more that an integration test in general doesn't seem warranted.

The `SchemaManager::createTable` and `SchemaManager::updateSchema` methods are the only two points of entry for DDL in the connector right now. If we can ensure that those methods invoke the proper BigQuery API methods with the correct schemas, that gives us fairly strong guarantees about the accuracy of the sanitization feature on the DDL front, and allows for much faster and more granular testing. If we don't want to engage in a lot of mocking and capturing of arguments passed to a mocked `BigQuery` instance, we can also directly test the `SchemaManager::getAndValidateProposedSchema` method.

The issue I see with the integration test is that none of the logic that is being tested actually requires interaction with BigQuery; it's all internal to the connector. And although it does give us end-to-end guarantees about one specific case, because of the time it takes to run a single test case, it becomes harder to add coverage for different permutations of properties (such as including the key or Kafka data fields, or enabling upsert/delete).

ncliang commented:
Fair enough. As I said, I added the IT more as a sanity check to verify the flow end-to-end. I agree that the runtime is prohibitive to adding more exhaustive testcases in IT. I've removed the IT and added more testcases in `SchemaManagerTest` . PTAL @C0urante

C0urante commented on kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/convert/KafkaDataBuilder.java:
```diff
@@ -52,7 +53,9 @@ public static Field buildKafkaDataField(String kafkaDataFieldName) {
         Field insertTimeField = insertTimeBuilder.build();
 
-        return Field.newBuilder(kafkaDataFieldName, LegacySQLTypeName.RECORD,
+        return Field.newBuilder(
+                kafkaDataFieldName,
+                LegacySQLTypeName.RECORD,
```
Can we revert this whitespace-only change? Might make a `pint merge` harder in the future.

C0urante reviewed (approved):
LGTM, thanks Nigel! One comment RE: a whitespace-only change but apart from that this is good to merge.

ncliang commented:
Merging as the failing `BigQueryErrorResponsesIT.testWriteToRecreatedTable` seems to already be failing on `2.0.x` branch before this change.
